src/parser/flow_lexer.ml

```diff
@@ -284,6 +292,107 @@ end = struct
       with No_good -> raise e
 end
 
+module IntOfString : sig
+  val int_of_string: string -> float
+end = struct
+  type t = {
+    negative: bool;
+    mantissa: int;
+    exponent: int;
+    todo: char list;
+  }
+
+  exception No_good
+
+  let eat f =
+    match f.todo with
+    | _::todo -> { f with todo; }
+    | _ -> raise No_good
+
+  let start str =
+    let todo = ref [] in
+    String.iter (fun c -> todo := c::(!todo)) str;
+    {
+      negative = false;
+      mantissa = 0;
+      exponent = 0;
+      todo = List.rev (!todo);
+    }
+
+  let parse_sign f =
+    match f.todo with
+    | '+'::_ -> eat f
+    | '-'::_ -> { (eat f) with negative = true; }
+    | _ -> f
+
+  let parse_hex_symbol f =
+    match f.todo with
+    | '0'::('x' | 'X')::_ -> f |> eat |> eat
+    | _ -> raise No_good
+
+  let parse_exponent f =
+    let todo_str = f.todo
+      |> List.map Char.escaped
+      |> String.concat "" in
+    let exponent =
+      try int_of_string todo_str
+      with Failure _ -> raise No_good in
+    { f with exponent; todo = [] }
+
+  let rec parse_body f =
+    match f.todo with
+    | [] -> f
+    (* _ is just ignored *)
+    | '_'::_ -> parse_body (eat f)
+    (* n is just ignored *)
+    | 'n'::_ -> parse_body (eat f)
+    | ('p' | 'P')::_ ->
+        parse_exponent (eat f)
+    | c::_ ->
+        let ref_char_code =
+          if c >= '0' && c <= '9'
+          then Char.code '0'
+          else if c >= 'A' && c <= 'F'
+          then Char.code 'A' - 10
+          else if c >= 'a' && c <= 'f'
+          then Char.code 'a' - 10
+          else raise No_good in
+        let value = (Char.code c) - ref_char_code in
+        let mantissa = (f.mantissa lsl 4) + value in
+        parse_body { (eat f) with mantissa; }
+
+  let float_of_t f =
+    assert (f.todo = []);
+    let ret = float_of_int f.mantissa in
+    let exponent = f.exponent in
+    let ret =
+      if exponent = 0
+      then ret
+      else ret ** (float_of_int exponent) in
+    if f.negative
+    then -.ret
+    else ret
+
+  let int_of_string str =
+    try Pervasives.float_of_string str
+    with e when Sys.win32 ->
+      try
+        start str
+          |> parse_sign
+          |> parse_hex_symbol
+          |> parse_body
+          |> float_of_t
+      with No_good -> raise e
+end
+
+let bigint_strip_n raw =
+  let size = String.length raw in
+  let str =
+    if (raw.[size - 1]) == 'n'
```

Comment:
If `size` is `0` then this is an exception. Not sure when `raw` is the empty string, but FYI

Reply:
I think it is never empty string, but should I check for `size == 0` just in case?

Reply:
So the real question here is "what invariants can `bigint_strip_n` assume about its input?"

The basic choices are:

1. Make no assumptions. Handle empty strings and strings without `n` at the end. Maybe throw in those cases, since the lexer should never generate these bad strings?
2. Assume a well-formed bigint string. No need to check for empty string and can assume it ends with `n`.

Right now it's kind of half & half. I prefer (1), but either is probably fine.

Reply:
Empty string would error in `Int64.of_string` 😄then
```ocaml
utop # Int64.of_string "";;
Exception: Failure "int_of_string".
utop # Pervasives.float_of_string "";;
Exception: Failure "float_of_string".
```